src/gradualizer_lib.erl

```diff
@@ -109,6 +112,8 @@ pick_value(?type(tuple, any)) ->
     {};
 pick_value(?type(tuple, Tys)) ->
     list_to_tuple([pick_value(Ty) || Ty <- Tys]);
+pick_value(Rec = ?type(record, _)) ->
+    prettyprint(Rec);
```

Comment:
If this function is not going to pick a value of the given type anymore, you better change its name and change all the other clauses too.

I wouldn't mind changing the error message into something like `"The patterns are non-exhaustive. Values of type #{foo :: integer()} are not covered by any clause"` instead of giving an example *value* which is not covered.

I still don't see why you can't actually pick a record value of the given record type, as is done for all the other types. To do that, you would create a record value like `{record, erl_anno:new(0), RecName, Fields}` where Fields is created using pick_value for each field type in the record type.

Reply:
Given that the values returned by other function clause are not using Erlang's abstract format (hopefully I'm using the right word), it makes little sense for the `record` type clause to be the only one returning a value compatible with the abstract format.

Which solution do we want to take?

1. Return all values as Erlang's abstract format (`pick_value(abstract_type()) -> [erl_parse:abstract_expr()]`) like the proposed example.

2. Return a formatted string (`pick_value(abstract_type()) -> [iolist()]`)

3. ?

Reply:
Ahh, I understand the problem now! Because pick_value returns a simple value, it is not possible to tell a record apart from a tuple, so if we return `#foo{bar = c}` here, it'd be printed as e.g. `{foo, c}`. At first, I didn't even realize it's a problem at all. :facepalm: 

Well then, spontaneously I'm leaning towards (1) since we're working with abstract syntax everywhere (abstract format, parse tree representation, etc. all these terms mean the same).

I don't prefer (2) because it means that we would have to do formatting in every clause. Doing it in one place seems better. It makes it uniform and easier to extend. It's already possible to provide a formatting fun to produce e.g. Elixir code in the error messages, but it's not applied in this particular error. I guess it should in the future (not in scope of this PR).

Another idea (3) could be to just include the type in the error message. Then we don't need pick_value. I gave an example in another comment the other day. We'd have to think about if the error messages will be easy to understand for each of the types, but I think they would. @josefs any opinion?

    test/should_fail/record_exhaustive.erl: Nonexhaustive patterns on line 47 at column 1
    Example values which are not covered: #union_rec{field = c}

vs

    test/should_fail/record_exhaustive.erl: Nonexhaustive patterns on line 47 at column 1
    Type of values which are not covered: #union_rec{field :: c | d | e}

Reply:
Thanks for laying out the options so well, @zuiderkwast .

The way I see it is that printing the type gives maximal information, but might be confusing because the types can look very contrived after refinement. Also, these types don't actually occurr anywhere in the program which adds to the confusion. Providing a concrete value is, at least in my mind, much simpler and easier to understand. It has the problem that we're not providing the full story. 

Using example values in the error messages optimizes for understandability. 

Other languages also tent to print values when reporting exhaustiveness errors, although they might try to print multiple values e.g. [ocaml](https://stackoverflow.com/questions/50801956/fatal-error-exception-match-failuremain-ml-8-15), [haskell](https://stackoverflow.com/questions/54726077/avoid-inappropriate-non-exhaustive-pattern-match-warning-in-ghci). 

Another option (4) would be to print both the refined type and an example value of that type. That would provide maximal information while at the same time provide an example to help understanding the type. A potential downside is information overload in the error message.

My preference is for (1) and (4), I'd like to keep the example values in the error. Of the two, I think I slightly prefer (4) since it offers all the information available. I agree that (2) should be avoided.

Reply:
OK, good points @josefs. So the natural incremental steps would be to first do (1) i.e. make pick_value return values in AST form and then later (4) also print the type. So for this PR, maybe only do (1)?

Reply:
Yes, that sounds reasonable.

Reply:
Sorry for the silence, I worked on other project during the holidays.

`pick_value` now returns a value in abstract format that is prettyprinted by `erl_pp` in `format_type_error` (solution 1)